autorest/azure/token.go

```diff
@@ -70,29 +77,72 @@ func (t *Token) WithAuthorization() autorest.PrepareDecorator {
 type ServicePrincipalToken struct {
 	Token
 
+	secret        ServicePrincipalSecret
 	clientID      string
-	clientSecret  string
-	resource      string
 	tenantID      string
+	resource      string
 	autoRefresh   bool
 	refreshWithin time.Duration
 	sender        autorest.Sender
 }
 
-// NewServicePrincipalToken creates a ServicePrincipalToken from the supplied Service Principal
-// credentials scoped to the named resource.
-func NewServicePrincipalToken(id string, secret string, tenantID string, resource string) (*ServicePrincipalToken, error) {
+// ServicePrincipalSecret is an interface that allows various secret mechanism to fill the form
+// that is submitted when acquiring an oAuth token
+type ServicePrincipalSecret interface {
```

Comment:
I prefer a top-down patten. Please move the types upon which ServicePrincipalToken depends before its definition.